### PR TITLE
Make date-time regexp RFC3339 compatible

### PR DESCRIPTION
**Pull request by Skeeve**
RFC3339 states that the UTC offset needs to have 2 digit hour and minute offset.

Previous regexp allows for 1+ digits so "+1:1" and "-0123456789:9876543210" would be accepted.

### Summary
While using Openapi Plugin I noticed that it doesn't accept ISO timestamps. While this is fine, as Openapi spec states that the format be RFC3339, I noticed that the regexp checking the dates is too lax.

### Motivation
I see no advantage in accepting wrong time offsets like 123:567. 

### References
Discussion on IRC. Don't know how to link to it?

### CODE REVIEW

**Review by jhthorsen — CHANGES_REQUESTED** [review]
Please add links to the specification and new tests.

**Comment by Skeeve** [comment]
Tests submitted.

[RFC3339](https://datatracker.ietf.org/doc/rfc3339/?include_text=1)

See:
```
   time-hour       = 2DIGIT  ; 00-23
   time-minute     = 2DIGIT  ; 00-59
   time-numoffset  = ("+" / "-") time-hour ":" time-minute
```

**Comment by Skeeve** [comment]
I noticed a wrong test in. you had a test against. `'2017-03-29\t23:02:55-12:00'` while you realy meant `"2017-03-29\t23:02:55-12:00"` as you didn want to check the non-acceptance of a `\` followed vy `t` but the non-acceptance of a tab. Fixed that.

Also a check that the time offset is in the range of 00:00 to 23:59.

File: lib/JSON/Validator/Formats.pm [hunk 1]
```diff
@@ -31,8 +31,10 @@ sub check_date {
 }
 
 sub check_date_time {
-  my @dt = $_[0] =~ m!^(\d{4})-(\d\d)-(\d\d)[T ](\d\d):(\d\d):(\d\d(?:\.\d+)?)(?:Z|([+-])(\d+):(\d+))?$!io;
+  my @dt = $_[0] =~ m!^(\d{4})-(\d\d)-(\d\d)[T ](\d\d):(\d\d):(\d\d(?:\.\d+)?)(?:Z|([+-])(\d\d):(\d\d))?$!io;
```
Comment (abraxxa) [inline lib/JSON/Validator/Formats.pm:34]: I'd prefer \d{2} for readability

Comment (Skeeve) [reply lib/JSON/Validator/Formats.pm:34]: \d\d was there in the first place and, tbh, I prefer \d\d for readability and only start \d{n} when n>2.

File: lib/JSON/Validator/Formats.pm [hunk 2]
```diff
@@ -31,8 +31,10 @@ sub check_date {
...
   return 'Does not match date-time format.' unless @dt;
+  return 'Time offset hour out of range.' if defined $dt[7] and $dt[7] > 23;
+  return 'Time offset minute out of range.' if defined $dt[8] and $dt[8] > 59;
```
Comment (jhthorsen) [inline lib/JSON/Validator/Formats.pm:37]: I don't understand why this is added. Isn't already covered in https://github.com/mojolicious/json-validator/blob/master/t/jv-formats.t#L29-L30 ?

File: t/jv-formats.t [hunk 3]
```diff
@@ -1,4 +1,5 @@
 use lib '.';
+use lib './lib';
```
Comment (jhthorsen) [inline t/jv-formats.t:2]: Please remove this.

**Review by jhthorsen — COMMENTED** [review]
There's a lot going on in this PR. Please squash your commits and make it to the point. I was thinking the PR would be something like this:

```perl
diff --git a/lib/JSON/Validator/Formats.pm b/lib/JSON/Validator/Formats.pm
index 4886fe5e..c91e43a1 100644
--- a/lib/JSON/Validator/Formats.pm
+++ b/lib/JSON/Validator/Formats.pm
@@ -31,7 +31,7 @@ sub check_date {
 }

 sub check_date_time {
-  my @dt = $_[0] =~ m!^(\d{4})-(\d\d)-(\d\d)[T ](\d\d):(\d\d):(\d\d(?:\.\d+)?)(?:Z|([+-])(\d+):(\d+))?$!io;
+  my @dt = $_[0] =~ m!^(\d{4})-(\d\d)-(\d\d)[T ](\d\d):(\d\d):(\d\d(?:\.\d+)?)(?:Z|([+-])(\d\d):(\d\d))?$!io;
   return 'Does not match date-time format.' unless @dt;
   @dt = map { s/^0//; $_ } reverse @dt[0 .. 5];
   $dt[4] -= 1;    # month are zero based
diff --git a/t/jv-formats.t b/t/jv-formats.t
index 24840305..b7e8a610 100644
--- a/t/jv-formats.t
+++ b/t/jv-formats.t
@@ -34,11 +34,21 @@ my $schema = {type => 'object', properties => {v => {type => 'string'}}};
   validate_ok {v => '2017-03-00T23:02:55Z'},       $schema, E('/v', 'Day out of range.');
   validate_ok {v => '2017-13-29T23:02:55Z'},       $schema, E('/v', 'Month out of range.');
   validate_ok {v => '2017-00-29T23:02:55Z'},       $schema, E('/v', 'Month out of range.');
+
+  # timezone offset
+  validate_ok {v => '2020-01-29T22:02:55-09:00'},   $schema;
+  validate_ok {v => '2020-01-29T22:02:55+06:30'},   $schema;
+  validate_ok {v => '2020-01-29T22:02:55-09'},      $schema, E('/v', 'Does not match date-time format.');
+  validate_ok {v => '2020-01-29T22:02:55+06'},      $schema, E('/v', 'Does not match date-time format.');
+  validate_ok {v => '2020-01-29T22:02:55-1234'},    $schema, E('/v', 'Does not match date-time format.');
+  validate_ok {v => '2020-01-29T22:02:55-123:456'}, $schema, E('/v', 'Does not match date-time format.');
+  validate_ok {v => '2020-01-29T22:02:55+1234'},    $schema, E('/v', 'Does not match date-time format.');
+  validate_ok {v => '2020-01-29T22:02:55+123:456'}, $schema, E('/v', 'Does not match date-time format.');
 }
```

File: lib/JSON/Validator/Formats.pm (hunk 2 above)
Comment (jhthorsen) [inline lib/JSON/Validator/Formats.pm:37]: Has this been run through perltidy?

File: t/jv-formats.t [hunk 4]
```diff
@@ -23,17 +24,22 @@ my $schema = {type => 'object', properties => {v => {type => 'string'}}};
...
+  validate_ok {v => '2017-03-29T23:02:55+23:60'},   $schema, E('/v', 'Time offset minute out of range.');
+  validate_ok {v => '2017-03-29T23:02:60Z'},        $schema, E('/v', 'Second out of range.');
+  validate_ok {v => '2017-03-29T23:61:55Z'},        $schema, E('/v', 'Minute out of range.');
+  validate_ok {v => '2017-03-29T24:02:55Z'},        $schema, E('/v', 'Hour out of range.');
+  validate_ok {v => '2017-03-32T23:02:55Z'},        $schema, E('/v', 'Day out of range.');
+  validate_ok {v => '2017-02-30T23:02:55Z'},        $schema, E('/v', 'Day out of range.');
+  validate_ok {v => '2017-02-29T23:02:55Z'},        $schema, E('/v', 'Day out of range.');
+  validate_ok {v => '2017-03-00T23:02:55Z'},        $schema, E('/v', 'Day out of range.');
+  validate_ok {v => '2017-13-29T23:02:55Z'},        $schema, E('/v', 'Month out of range.');
+  validate_ok {v => '2017-00-29T23:02:55Z'},        $schema, E('/v', 'Month out of range.');
```
Comment (jhthorsen) [inline t/jv-formats.t:42]: Please add some "ok" tests for the time offset.

File: t/jv-formats.t (hunk 3 above)
Comment (Skeeve) [reply t/jv-formats.t:2]: removed

File: lib/JSON/Validator/Formats.pm (hunk 2 above)
Comment (Skeeve) [reply lib/JSON/Validator/Formats.pm:37]: Now it's done

File: t/jv-formats.t (hunk 4 above)
Comment (Skeeve) [reply t/jv-formats.t:42]: There already are some at the begin of the tests

https://github.com/mojolicious/json-validator/blob/bf879a91c804423be29adf722a05c03f925ac8b7/t/jv-formats.t#L29

I added two more with offsets +23:59 and -23:59

**Comment by Skeeve** [comment]
Since I messed it up, see a clean version here https://github.com/mojolicious/json-validator/pull/238

**Pull request closed without merging**